racket/collects/racket/stream.rkt

```diff
@@ -179,3 +184,129 @@
          (cond [(stream-empty? s) empty-stream]
                [else (stream-cons e (stream-cons (stream-first s)
                                                  (loop (stream-rest s))))])))))
+
+;; Impersonators and Chaperones ----------------------------------------------------------------------
+;; (these are private because they would fail on lists, which satisfy `stream?`)
+
+(define (impersonate-stream s first-proc rest-proc . props)
+  (impersonate-generics
+   gen:stream
+   ; the struct-info param is a hack, see PR 14970
+   (apply impersonate-struct s struct-info (λ (a b) (values a b)) props)
+   [stream-first
+    (λ (stream-first)
+      (impersonate-procedure stream-first
+                             (λ (s) (values (λ (v) (first-proc v)) s))))]
+   [stream-rest
+    (λ (stream-rest)
+      (impersonate-procedure stream-rest
+                             (λ (s) (values (λ (v) (rest-proc v)) s))))]))
+
+(define (chaperone-stream s first-proc rest-proc . props)
+  (chaperone-generics
+   gen:stream
+   ; the struct-info param is a hack, see PR 14970
+   (apply chaperone-struct s struct-info (λ (a b) (values a b)) props)
+   [stream-first
+    (λ (stream-first)
+      (chaperone-procedure stream-first
+                           (λ (s) (values (λ (v) (first-proc v)) s))))]
+   [stream-rest
+    (λ (stream-rest)
+      (chaperone-procedure stream-rest
+                           (λ (s) (values (λ (v) (rest-proc v)) s))))]))
+
+;; Stream contracts ----------------------------------------------------------------------------------
+
+(define (stream/c-name ctc)
+  (define elem-name (contract-name (base-stream/c-content ctc)))
+  (apply build-compound-type-name
+         'stream/c
+         elem-name
+         '()))
+
+(define (add-stream-context blame)
+  (blame-add-context blame "a value generated by"))
+
+(define (check-stream/c ctc val blame)
+  (unless (stream? val)
+    (raise-blame-error blame val '(expected "a stream" given: "~e") val)))
+
+(define (check-stream/c-np ctc val blame)
+  (if (stream? val)
+      #f
+      (λ (neg-party)
+        (raise-blame-error blame #:missing-party neg-party
+                           val '(expected "a stream" given: "~e") val))))
+
+(define (stream/c-stronger? a b)
+  (contract-stronger? (base-stream/c-content a) (base-stream/c-content b)))
+
+; streams are lazy, so we need to contract the rest of the stream lazily (which can be a list)
+(define (contract-stream-rest v ctc blame)
+  (define elem-ctc (base-stream/c-content ctc))
+  (define new-ctc (if (list? v) (listof elem-ctc) ctc))
+  (contract new-ctc v
```

Comment:
@rfindler Unless I'm missing something about how contracts work, I think that attaching a new contract is necessary, since streams are lazy, and different elements of the stream are fundamentally different objects.

Because of that, contracts need to be assigned both *lazily* and deeply. Therefore, `(stream-rest s)` needs to have a new contract attached to it so that further calls to `stream-rest` on the same stream still include the contract.